momento-provider/src/main/java/com/google/code/ssm/providers/momento/MomentoClientWrapper.java

```diff
@@ -217,11 +196,67 @@ public Object getNativeClient() {
         return transcoderAdapter;
     }
 
-    private static int getFlag(byte[] data, int i) {
-        return (data[i] & 0xff) << 24
-                | (data[i + 1] & 0xff) << 16
-                | (data[i + 2] & 0xff) << 8
-                | (data[i + 3] & 0xff);
+    // Helper function that concats the CachedObject to raw bytes before calling `set`.
+    // This helps maintain our serialization flags so that we can deserialize the object upon a `get`
+    private boolean writeOutToMomento(final String key, final int exp, final int flagsUsed, final byte[] data) {
+        final byte[] flags = intToBytes(flagsUsed);
+        ByteBuffer buffer = ByteBuffer.allocate(flags.length + data.length);
+        buffer.put(flags);
+        buffer.put(data);
+        buffer.rewind();
+        CacheSetResponse response = momentoClient.set(defaultCacheName, key, buffer.asReadOnlyBuffer(), exp);
+        return response != null;
+    }
+
+    // Helper function that reads the raw bytes from Momento and splits up the bytes
+    // to extract the serialization flags we used as well as the raw bytes of data that contain our
+    // actual object
+    private <T> T readFromMomento(final String key, final CacheTranscoder transcoder) {
+        Transcoder<T> cacheTranscoder = getTranscoder(transcoder);
+        CacheGetResponse response = momentoClient.get(defaultCacheName, key);
+        Optional<byte[]> cacheGetResponse = response.byteArray();
+        if (cacheGetResponse.isPresent()) {
+            byte[] returnedBytes = cacheGetResponse.get();
+            int flags = getFlags(returnedBytes);
+            byte[] originalData = getData(returnedBytes);
+            return cacheTranscoder.decode(new CachedData(flags, originalData, originalData.length));
```

Comment:
does your transcoder not support ByteBuffer? This line ~implies~ directly means that the heap byte array memory cost to read an object from Momento is 2x the size of that object, before even deserializing it. If instead you replaced line 219-220 with 
```
ByteBuffer readBuffer = ByteBuffer.wrap(cacheGetResponse.get());
const int  flags = readBuffer.getInt();
// remove getData(), it copies the whole response payload on the heap just to move a deserialization cursor forward.
```
then the read buffer would be in a state where the flags are consumed and the data remains, with no extra heap copies of the whole buffer.

Reply:
++ This is excellent, this is definitely going to be more performant. Thank you!

Reply:
And to answer your question, no the transcoders don't support ByteBuffer, they wrap everything up into a `CachedObject`/`CachedData` construct

Reply:
that is incredible and truly awful.